readthedocs/oauth/services/gitlab.py

```diff
@@ -0,0 +1,222 @@
+
+            if repo.organization and repo.organization != organization:
+                log.debug('Not importing %s because mismatched orgs' %
+                          fields['name'])
+                return None
+            else:
+                repo.organization = organization
+            repo.name = fields['name']
+            repo.full_name = fields['name_with_namespace']
+            repo.description = fields['description']
+            repo.ssh_url = fields['ssh_url_to_repo']
+            repo.html_url = fields['web_url']
+            repo.private = not fields['public']
+            repo.clone_url = fields['http_url_to_repo']
+            repo.admin = not repo_is_public
+            if not repo.admin and organization:
+                repo.admin = is_owned_by(fields['owner']['id'])
+            repo.vcs = 'git'
+            repo.account = self.account
+            repo.avatar_url = fields.get('avatar_url')
+            repo.json = json.dumps(fields)
+            repo.save()
+            return repo
+        else:
+            log.info(
+                'Not importing {0} because mismatched type: public={1}'.format(
+                    fields['name_with_namespace'],
+                    fields['public'],
+                )
+            )
+
+    def create_organization(self, fields):
+        """Update or create remote organization from GitLab API response
+
+        :param fields: dictionary response of data from API
+        :rtype: RemoteOrganization
+        """
+        try:
+            organization = RemoteOrganization.objects.get(
+                slug=fields.get('path'),
+                users=self.user,
+                account=self.account,
+            )
+        except RemoteOrganization.DoesNotExist:
+            organization = RemoteOrganization.objects.create(
+                slug=fields.get('path'),
+                account=self.account,
+            )
+            organization.users.add(self.user)
+
+        organization.name = fields.get('name')
+        organization.account = self.account
+        organization.url = '{url}/{path}'.format(
+            url=self.adapter.provider_base_url, path=fields.get('path')
+        )
+        if fields.get('avatar'):
+            organization.avatar_url = '{url}/{avatar}'.format(
+                url=self.adapter.provider_base_url,
+                avatar=fields['avatar']['url'],
+            )
+        organization.json = json.dumps(fields)
+        organization.save()
+        return organization
+
+    def setup_webhook(self, project):
+        """Set up GitLab project webhook for project
+
+        :param project: project to set up webhook for
+        :type project: Project
+        :returns: boolean based on webhook set up success
+        :rtype: bool
+        """
+        session = self.get_session()
+
+        # See: http://doc.gitlab.com/ce/api/projects.html#add-project-hook
+        data = json.dumps({
+            'id': 'readthedocs',
```

Comment:
indeed, i assume this would fail or at least overwrite the existing webhook if a user set up a webhook with readthedocs.org and readthedocs.com -- or a private instance.